openedx_tagging/core/tagging/models.py

```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+
+    On Tag.external_id stores an identifier from the instance (`pk` as default)
+    and on Tag.value stores an human readable representation of the instance
+    (e.g. `username`). 
+    The subclasses can override this behavior, to choose the right field.
+
+    When an ObjectTag is created with an existing Tag, 
+    the Tag is re-synchronized with its instance.
+    """
+
+    class Meta:
+        proxy = True
+
+    def __init__(
+        self,
+        *args: Any,
+        **kwargs: Any
+    ) -> None:
+        """
+        Checks if the `tag_class_model` is correct
+        """
+        assert issubclass(self.tag_class_model, models.Model)
+        super().__init__(*args, **kwargs)
+
+    @property
+    def tag_class_model(self) -> Type:
+        """
+        Subclasses must implement this method to return the Django.model
+        class referenced by these object tags.
+        """
+        raise NotImplementedError
+    
+    def _check_instance(
+        self,
+        object_tag: ObjectTag
+    ) -> bool:
+        """
+        Returns True if the instance exists
+
+        Subclasses can override this method to perform their own instance validation checks.
+        """
+        intance_id = object_tag.tag.external_id
+        return self.tag_class_model.objects.filter(pk=intance_id).exists()
+
+    def _check_tag(
+        self,
+        object_tag: ObjectTag
+    ) -> bool:
+        """
+        Returns True if pass the instance validation checks
+        """
+        return (
+            super()._check_tag(object_tag)
+            and self._check_instance(object_tag)
+        )
+    
+    def _get_external_id_from_instance(
+        self,
+        instance
+    ) -> str:
+        """
+        Returns the tag external id from the instance
+
+        Subclasses can override this method to get the external id from their own models.
+        """
+        return str(instance.pk)
+
+    def _get_value_from_instance(
+        self,
+        instance
+    ) -> str:
+        """
+        Returns the tag value from the instance
+
+        Subclasses can override this method to get the value from their own models.
+        """
+        return str(instance.pk)
+    
+    def _get_instance(self, pk):
+        """
+        Gets the instance from `pk`
+        """
+        try:
+            return self.tag_class_model.objects.get(pk=pk)
+        except self.tag_class_model.DoesNotExist:
+            return None
+    
+    def _resync_tag(self, tag: Tag) -> Tag:
+        """
+        Resync the tag value with the value from the instance
+        """
+        instance = self._get_instance(tag.external_id)
+        if instance:
+            value = self._get_value_from_instance(instance)
+            if tag.value != value:
+                tag.value = value
+                tag.save()
+        return tag
+
+    def _set_tag(
+        self,
+        tag_ref,
+        object_tag: ObjectTag
+    ) -> ObjectTag:
```

Comment:
I created a [property setter for `ObjectTag.tag_ref`](https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R545) which handles the dynamic Tag creation like you've done here.

It's [cleaner to call](https://github.com/open-craft/openedx-learning/pull/6/commits/9fd8854c293cc9b506c2a55181c5aab9f6b2b80d#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R404), and I think makes more sense from an encapsulation perspective.

Reply:
Added to the branch